src/pybind/mgr/mgr_util.py

```diff
@@ -66,3 +66,22 @@ def format_dimless(n, width, colored=True):
 
 def format_bytes(n, width, colored=True):
     return format_units(n, width, colored, decimal=False)
+
+
+def differentiate(data1, data2):
+    """
+    # >>> times = [0, 2]
+    # >>> values = [100, 101]
+    # >>> differentiate(*zip(times, values))
+    0.5
+    """
+    try:
+        return (data2[1] - data1[1]) / float(data2[0] - data1[0])
+    except ZeroDivisionError:
+        return 0.0
```

Comment:
Just to understand, when is this case happening? A possible alternative could be to return `float('nan')` (Not a Number) as assuming this is `0.0` can be misleading. `nan` is a perfectly valid Python float object, so you can still perform arithmetic operations with it:
```python
>>> float('nan') + 1
nan
>>> float('nan')*2
nan
>>> float('nan')/float('nan')
nan
```

Reply:
[JOSN](https://www.json.org/) left out NaN, but `json.dumps` generates them:

```python
In [2]: json.dumps(float('nan'))
Out[2]: 'NaN'
```
I'd like to keep it simple and not run into this complexity.

Reply:
IMHO `0.0` is not good because it gives a wrong output.

It seems that, while a lot of JSON parsers are capable of dealing with `NaN` (it's a valid JS `number` value), it's not [strict JSON](http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf):

> Numeric values that cannot be represented as sequences of digits (such as Infinity and NaN) are not permitted

(That also implies that we should use [`json.dumps(..., allow_nan=False)`](https://docs.python.org/2/library/json.html#json.dump) always if we want to comply with the ECMA JSON standard).

In that case you can leverage on the `null` value, as it's perfectly strict JSON. I would do the mapping 'NaN' to `null` in the JSON dumping, while keeping `nan` for other output formats.

Reply:
When returning something like `[0.4, 0.23, 123.0, 123.5, null, 42, 42,42, 0.5, 0.0]`, I wonder how many null related exceptions we're going to generate in client code. I don't think lots of users will expect `null` values here.

Reply:
I share that concern too, but the issue here is whether it's preferable something that _might not_ work:
```js
[0.4, 0.23, 123.0, 123.5, **null**, 42, 42,42, 0.5, 0.0]
```
Than something that works but contains false data:
```js
[0.4, 0.23, 123.0, 123.5, **0.0**, 42, 42,42, 0.5, 0.0]
```
If someone draws a chart from the latter data, they will see the graph suddenly dropping to zero. Most visualization engines (e.g.: Grafana) properly treat `null` values, but 0.0 is a perfectly valid value.

If you use Python to parse that, you'll get a `None` (Java, PHP, etc., do the same), which will quickly trigger an exception if you try to do math on it:
```python
>>> import json
>>> json.loads('[0.4, 0.23, 123.0, 123.5, null, 42, 42,42, 0.5, 0.0]')
[0.4, 0.23, 123.0, 123.5, None, 42, 42, 42, 0.5, 0.0]
```

Reply:
And yes, I also share your concern. Anyother way would be to amend `mgr_util.latest_rate` to look further ahead, in case the last two timestamps are identical.

Reply:
also: #26045

Reply:
I agree that returning `0.0` can be misleading (you can think that this was the original value)...
If the API tests are already covering the case for null values, we can go for it and run the tests to make sure that workflows are not broken.

Reply:
What's the current state of this PR? Is there a final decision how to handle this issue?

Reply:
@votdev yeah, the discussion made this a bit more complicated that I originally anticipated. I'm hesitant to return `null`s here, as this is something unexpected for users (external ones) of our API.

Reply:
@sebastian-philipp don't want to become a burden for this PR, so if anyone else is happy with returning `0.0` for `NaN`s, no objections from my side :wink:

Reply:
so, the big problem right now is: you're right.

Reply:
Haha, ok then let's step back and see: who's the consumer of this? Just our Angular services, which basically rely on the browser's JSON parser, which AFAIK all of them support JSON null objects, so the issue then becomes how our Angular code deals with nulls. Well, (un)fortunately JS/TS seem to be extremely friendly when it comes to coping with null/undefined variables (no segfaults as in C++). That makes the issue hard to catch, but at the same time less critical. IMHO, let's move on and see: the quickest way find if a pipe leaks is to pump water into it.

Reply:
I'm more tending to return ``NaN`` because this indicates that there was an error, returning 0.0 is a incorrect and false data here. But if 0.0 simplifies the processing in an upper layer, then i am also happy with this.

Reply:
Well, this PR touches all mgr modules, not only the dashboard module. And adding stray `NaN`s or `None`s, is not something I want to introduce.

Reply:
Well, adding stray and suprising `null`s lead to bugs like #26766

Reply:
> Well, adding stray and surprising `null`s lead to bugs like #26766

@sebastian-philipp @LenzGr we can still go with null values and target this fix for master (not nautilus) and then fix the not handled errors found. Thoughts?

Reply:
this is a function that affects **all** mgr modules, not just the dashboard.

Reply:
> @sebastian-philipp: Well, adding stray and surprising nulls lead to bugs like #26766

And I think that #26766 does the proper thing: handling all possible inputs and not mapping `nulls` or `undefined` to `0`, which are semantically different values.

> @sebastian-philipp: this is a function that affects all mgr modules, not just the dashboard.

I think that, besides dashboard, there are just 2 other modules using that: `diskprediction_cloud` and `status`, so let's just ping their top contributors for their feedback on the impact of this change: @hsiang41, @gmayyyha and @liewegas.

Reply:
> > @sebastian-philipp: Well, adding stray and surprising nulls lead to bugs like #26766
> 
> And I think that #26766 does the proper thing: handling all possible inputs and not mapping `nulls` or `undefined` to `0`, which are semantically different values.

I think so, too. I plan to amend this PR this week. 
> 
> > @sebastian-philipp: this is a function that affects all mgr modules, not just the dashboard.
> 
> I think that, besides dashboard, there are just 2 other modules using that: `diskprediction_cloud` and `status`, so let's just ping their top contributors for their feedback on the impact of this change: @hsiang41, @gmayyyha and @liewegas.

In case the dashboard becomes the official API, we should make sure, we're not returning nasty surprises, like stray `null` values. But yep, let's huddle if it makes sense.

Reply:
Sorry for coming late to this discussion. 
I'm going to step back a bit on the discussion of whether we should return `0.0` or `None` when we get a `ZeroDivisionError` exception, and analyze why are we getting a `ZeroDivisionError` when calculating the derivative of a sequence of data points.

If we are getting a `ZeroDivisionError` is because we have two consecutive data points that have two (possibly different) values for the same timestamp. In mathematical terms, this is the same as saying that `f(t) = a and f(t) = b` at the same time, which is impossible. When calculating a derivative of a sequence of data points we are assuming that the data points were generated by a continuous function, and thus the derivative function expects that `data2[0] - data1[0]` will be always different from zero.
Since this is not the case because of timestamp resolution, we might get two perf counter values for the same timestamp. So what we should do is to filter those consecutive data points that have the same timestamp and keep just one of them. 

Now we need to decide on the filtering algorithm for these special datapoints. Given a sequence of data points with the same timestamp, we can:
a) keep just the first data point of the sequence
b) keep just the last data point of the sequence
c) calculate the average of the data points

By doing this filtering we will never have a `ZeroDivisionError` again while calculating the derivative function, and solves this problem.

Reply:
IMO we should just calculate the difference (option c), but using the absolute like I did in my draft PR #28603 . Just calculate the time difference first and use `1` if it's `0`.

```
    td = float(data2[0] - data1[0])
    return abs((data2[1] - data1[1]) / (td if td != 0 else 1))
```

Reply:
I agree here. Why do we get those broken perf values from C++ in the first place?

Reply:
> I agree here. Why do we get those broken perf values from C++ in the first place?

It's not broken. The problem is that two perf values could have been separated by only a few nanoseconds but the clock only had milisecond resolution.